src/react/components.js

```diff
@@ -273,15 +275,85 @@ export function applyGetDerivedStateFromProps(
   invariant(getDerivedStateFromPropsCall !== undefined);
   let partialState = getDerivedStateFromPropsCall(realm.intrinsics.null, [props, prevState]);
 
-  if (partialState !== realm.intrinsics.null && partialState !== realm.intrinsics.undefined) {
+  const deriveState = state => {
     let objectAssign = Get(realm, realm.intrinsics.Object, "assign");
     invariant(objectAssign instanceof ECMAScriptFunctionValue);
     let objectAssignCall = objectAssign.$Call;
     invariant(objectAssignCall !== undefined);
-    let newState = new ObjectValue(realm, realm.intrinsics.ObjectPrototype);
-    invariant(objectAssignCall !== undefined);
-    objectAssignCall(realm.intrinsics.undefined, [newState, prevState, partialState]);
 
+    if (state instanceof AbstractValue && !(state instanceof AbstractObjectValue)) {
+      const kind = state.kind;
+
+      if (kind === "conditional") {
+        let condition = state.args[0];
+        let a = deriveState(state.args[1]);
+        let b = deriveState(state.args[2]);
+        invariant(condition instanceof AbstractValue);
+        if (a === null && b === null) {
+          return null;
+        } else if (a === null) {
+          invariant(b instanceof Value);
+          return AbstractValue.createFromConditionalOp(realm, condition, realm.intrinsics.false, b);
```

Comment:
Returning `false` here seems odd. Why not just `null`?

Reply:
`false` is better to minify (`!1`) and I believe it has better runtime costs for boolean casting in non JIT environments.